relay-server/src/services/project.rs

```diff
@@ -1137,102 +1138,79 @@ impl Project {
-                    "there is no project state: dropping {len} buckets"
-                );
-            }
-            CheckedBuckets::ProjectDisabled(len) => {
-                relay_log::debug!("dropping {len} buckets for disabled project");
-            }
-
-            CheckedBuckets::NoScoping(len) => {
-                relay_log::error!(
-                    tags.project_key = self.project_key.as_str(),
-                    "there is no scoping: dropping {len} buckets"
-                );
-            }
-            CheckedBuckets::RateLimited(len) => {
-                relay_log::debug!("dropping {len} buckets due to rate limit");
-            }
-        };
-
-        None
-    }
-
-    /// Unit-testable helper function for [`Self::check_buckets`].
-    ///
-    /// Returns [`CheckedBuckets::Ok`] if metrics are currently allowed, or the reject reason otherwise.
-    fn check_buckets_inner(
-        &mut self,
-        outcome_aggregator: Addr<TrackOutcome>,
-        buckets: Vec<Bucket>,
-    ) -> CheckedBuckets {
-        let len = buckets.len();
-
         let Some(project_state) = self.valid_state() else {
-            return CheckedBuckets::ProjectExpired(len);
+            relay_log::error!(
+                tags.project_key = self.project_key.as_str(),
+                "there is no project state: dropping {} buckets",
+                buckets.len(),
+            );
+            return None;
         };
 
         if project_state.invalid() || project_state.disabled() {
-            return CheckedBuckets::ProjectDisabled(len);
+            relay_log::debug!("dropping {} buckets for disabled project", buckets.len());
+            return None;
         }
 
         let Some(scoping) = self.scoping() else {
-            return CheckedBuckets::NoScoping(len);
+            relay_log::error!(
+                tags.project_key = self.project_key.as_str(),
+                "there is no scoping: dropping {} buckets",
+                buckets.len(),
+            );
+            return None;
         };
 
-        let limits = self.rate_limits().check_with_quotas(
-            project_state.get_quotas(),
-            scoping.item(DataCategory::MetricBucket),
-        );
+        let namespaces: BTreeSet<MetricNamespace> = buckets
+            .iter()
+            .filter_map(|bucket| bucket.name.try_namespace())
+            .collect();
 
-        if limits.is_limited() {
-            let mode = project_state.get_extraction_mode();
-            let reason_code = limits.longest().and_then(|limit| limit.reason_code.clone());
-            utils::reject_metrics::<Vec<Bucket>>(
-                &outcome_aggregator,
-                utils::extract_metric_quantities(&buckets, mode),
-                scoping,
-                Outcome::RateLimited(reason_code),
-                None,
-                None,
+        let mode = project_state.get_extraction_mode();
+        for namespace in namespaces {
```

Comment:
I think we can check the limits for all namespaces, collect the namespaces which are limited then extract quantities and drop buckets only once.

Reply:
Right. I had that at first and discarded it as it caused much more code and required yet another map with quantities along with refactors to SourceQuantities to allow increments. That added code didn't seem worth the minor performance improvement (if any).

Reply:
Ah you're right for a second I thought we didn't need to know which limit was triggered just that a limit for that namespace triggered. But we do need it for the reason code otherwise we could just collect a single source quantities.